tensorflow_addons/optimizers/optimizer_test_base.py

```diff
@@ -0,0 +1,149 @@
+import tensorflow as tf
+
+
+class OptimizerTestBase(tf.test.TestCase):
+    """Base class for optimizer tests.
+
+    Optimizer tests may inherit from this class and define test
+    functions using doTest. Usually this should include the functions
+    testSparse, testBasic, and testBasicCallableParams. See
+    weight_decay_optimizers_test for an example.
+    """
+
+    def doTest(self, optimizer, update_fn, params, do_sparse=False):
+        """The major test function.
+
+        Args:
+            optimizer: The tensorflow optimizer class to be tested.
+            update_fn: The numpy update function of the optimizer, the function
+                signature must be
+                update_fn(var: np.array,
+                          grad_t: np.array,
+                          slot_vars: dict,
+                          optimizer_params: dict) -> (updated_var,
+                                                      updated_slot_vars)
+                Note that slot_vars will be initialized to an empty dictionary
+                for each variable, initial values should be handled in the
+                update_fn.
+            params: A dict, the parameters to pass to the construcor of the
+                optimizer. Either a constant or a callable. This also passed to
+                    the optimizer_params in the update_fn.
+            do_sparse: If True, test sparse update. Defaults to False, i.e.,
+                dense update.
+        """
+        for i, dtype in enumerate([tf.half, tf.float32, tf.float64]):
+            with self.session(graph=tf.Graph()):
+                # Initialize variables for numpy implementation.
+                np_slot_vars0, np_slot_vars1 = {}, {}
+                var0_np = np.array([1.0, 2.0], dtype=dtype.as_numpy_dtype)
+                grads0_np = np.array([0.1, 0.1], dtype=dtype.as_numpy_dtype)
+                var1_np = np.array([3.0, 4.0], dtype=dtype.as_numpy_dtype)
+                grads1_np = np.array([0.01, 0.01], dtype=dtype.as_numpy_dtype)
+                # Create Tensorflow variables.
+                var0 = tf.Variable(var0_np, name="var0_%d" % i)
+                var1 = tf.Variable(var1_np, name="var1_%d" % i)
+                if do_sparse:
+                    grads0_np_indices = np.array([0, 1], dtype=np.int32)
+                    grads0 = tf.IndexedSlices(
+                        tf.constant(grads0_np), tf.constant(grads0_np_indices),
+                        tf.constant([2]))
+                    grads1_np_indices = np.array([0, 1], dtype=np.int32)
+                    grads1 = tf.IndexedSlices(
+                        tf.constant(grads1_np), tf.constant(grads1_np_indices),
+                        tf.constant([2]))
+                else:
+                    grads0 = tf.constant(grads0_np)
+                    grads1 = tf.constant(grads1_np)
+                opt = optimizer(**params)
+                # Validate initial values.
+                if not tf.executing_eagerly():
```

Comment:
Do we still need it?

Reply:
If we remove `test_utils.run_in_graph_and_eager_modes` and only run eager tests then we can remove the graph tests. I wasn't sure what the current best practice is and to what extent graph mode tests are still required due to `tf.function`?

Reply:
I agree, perhaps we can leave a TODO comment here, remind us to remove it later. What do you think?

Reply:
I'm fine with a TODO but these tend to stay longer than often intended ;) Do you know who might be able to shed some light on this?

Reply:
@karmel Do you have any thoughts about this. We've been kind of all over on required tests. If we do not test with `run_in_graph_and_eager_modes` we can't support compat.v1 which is not ideal imo, but we could make that a standard.

If we do remove compat.v1 tests, should we have another util that tests everything in eager  / decorated with `tf.function`? Some of our functions are already decorated with `tf.function`, but for those that are not... there is no guarantee that a user won't add it themselves.

Reply:
I change my mind after carefully reading those codes, I think they are valid because we shouldn't assume that optimizer runs inside a `tf.function` function.